### DISCUSSION TITLE
IIncrementalGenerator Database Examples?

### DISCUSSION THREAD

**Discussion by jw56578** [dotnet/roslyn#70887] [Q&A] [answered]
Is it possible to use SqlClient to query sql server using IIncrementalGenerator.
I have coded a prototype, but when I introduce SqlClient code into RegisterSourceOutput, it either throws an exception 
Could not load file or assembly System.Data.SqlClient. 
or the method is not called when debugging and the source code is never generated.
if I remove all reference to SqlClient code. It runs fine.

**Comment by jasonmalinowski** [comment]
Source generators should be quite fast and not have external dependencies; you should not be doing calls to external servers like SQL.

**Comment by sharwell** [comment, marked as answer]
External I/O (including but not limited to reading or writing files on disk) is not allowed in a source generator. All non-source inputs to a source generator need to be gathered during earlier steps of the build, and provided to the source generator through the `<AdditionalFiles>` mechanism in the compiler. A pre-build step to gather this information in a file could be designed to read data from an SQL server.
